deepchem/models/jax_models/pinns_model.py

```diff
@@ -0,0 +1,391 @@
+from deepchem.trans.transformers import Transformer, undo_transforms
+
+# JAX dependencies
+import jax
+import jax.numpy as jnp
+import haiku as hk
+import optax
+from deepchem.models.jax_models.jax_model import JaxModel
+from deepchem.models.jax_models.jax_model import create_default_gradient_fn, create_default_eval_fn
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+def create_default_update_fn(optimizer, model_loss):
+  """
+  This function calls the update function, to implement the backpropogation
+  """
+
+  @jax.jit
+  def update(params, opt_state, batch, target, weights,
+             rng) -> Tuple[hk.Params, optax.OptState, jnp.ndarray]:
+    batch_loss, grads = jax.value_and_grad(model_loss)(params, target, weights,
+                                                       rng, *batch)
+    updates, opt_state = optimizer.update(grads, opt_state)
+    new_params = optax.apply_updates(params, updates)
+    return new_params, opt_state, batch_loss
+
+  return update
+
+
+class PinnModel(JaxModel):
+  """
+  This is class is derived from the JaxModel class and methods are also very similar to JaxModel,
+  but it has the optional of passing multiple arguments(Done using *args) suitable for PINNs model.
+  Ex - Approximating f(x, y, z, t) satisfying a Linear diffrential equation.
+
+  This model is reccomended for Linear diffrential equations but if you can accurately write
+  the gradient function in Jax depending on your use case, then it will work as well.
+
+  This class requires two functions apart from the usual function definition and weights
+
+  [1] **grad_fn** : Each PINNs have a diffrent stratergy for calculating its final losses. This 
+  function tells the PinnModel on how to go about computing the derivatives for backpropogation.
+  It should follow this format:
+
+  >>>
+  >> def gradient_fn(forward_fn, loss_outputs, initial_data):
+  >>
+  >>  def model_loss(params, target, weights, rng, ...):
+  >>
+  >>    # write code using the arguments.
+  >>    # ... indicates the variable number of positional arguments.
+  >>    return
+  >>
+  >>  return model_loss
+
+  "..." can be replaced with various arguments like (x, y, z, y) but should match with eval_fn
+
+  [2] **eval_fn**: Function for defining on how the model needs to compute during inference.
+  It should follow this format
+
+  >>>
+  >> def create_eval_fn(forward_fn, params):
+  >>  def eval_model(..., rng=None):
+  >>    # write code here using arguments
+  >>
+  >>    return
+  >>  return eval_model
+
+  "..." can be replaced with various arguments like (x, y, z, y) but should match with grad_fn
+
+  [3] boundary_data:
+  For a detailed example, check out - deepchem/models/jax_models/tests/test_pinn.py where we have
+  solved f'(x) = -sin(x)
+
+  References
+  ----------
+  .. [1] Raissi et. al. "Physics Informed Deep Learning (Part I): Data-driven
+     Solutions of Nonlinear Partial Differential Equations" arXiv preprint arXiv:1711.10561
+
```

Comment:
You can also cite the original PINN paper: https://www.sciencedirect.com/science/article/pii/S0021999118307125